driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2272,12 +2272,14 @@ private Connection prepareAllQueries(Host host, Connection reusedConnection) thr
 
                 for (String keyspace : perKeyspace.keySet()) {
                     // Empty string mean no particular keyspace to set
-                    if (!keyspace.isEmpty())
+                    // Optimization: Only change keyspace for older protocol versions as newer protocols allow
+                    // specifying keyspace on prepared statement.
+                    if (protocolVersion().compareTo(ProtocolVersion.V5) < 0 && !keyspace.isEmpty())
                         connection.setKeyspace(keyspace);
 
                     List<Connection.Future> futures = new ArrayList<Connection.Future>(preparedQueries.size());
                     for (String query : perKeyspace.get(keyspace)) {
-                        futures.add(connection.write(new Requests.Prepare(query)));
+                        futures.add(connection.write(new Requests.Prepare(query, keyspace)));
```

Comment:
One thing here, since we can't distinguish whether the keyspace was established via the session keyspace or set on the statement explicitly we always pass it here.  I don't consider this that big of a deal due to the infrequency of prepared statement preparations.